File: problems/0230_kth-smallest-element-in-a-bst/solution_inorder_recursive.cpp
class Solution {
  public:
    int kthSmallest(TreeNode *root, int k) {
        // In-order traversal of a BST visits values in ascending order, so
        // the kth visit is the kth smallest. k is passed by reference and
        // counted down; answer records the kth visit's value.
        int answer = -1;
        // Recursion depth is bounded by the tree height h (worst case n on
        // a chain), which is why the iterative twin exists.
        inorder(root, k, answer);
        return answer;
    }

  private:
    void inorder(TreeNode *node, int &k, int &answer) {
        // Early stop: once the answer is recorded, the unvisited remainder
        // of the tree is never touched.
        if (node == nullptr || k == 0)
            return;
        inorder(node->left, k, answer);
        --k;
        if (k == 0) {
            answer = node->val;
            return;
        }
        inorder(node->right, k, answer);
    }
};

File: problems/0201-0300/0230_select-kth-smallest-in-a-search-tree/solution_inorder_recursive.java
class Solution {

    public int selectKthSmallest(TreeNode root, int k) {
        // In-order traversal of a BST visits values in ascending order, so
        // the kth visit is the kth smallest. k travels as state[0] and is
        // counted down; state[1] records the kth visit's value.
        int[] state = { k, -1 };
        // Recursion depth is bounded by the tree height h (worst case n on
        // a chain), which is why the iterative twin exists.
        inorder(root, state);
        return state[1];
    }

    private void inorder(TreeNode node, int[] state) {
        // Early stop: once the answer is recorded, the unvisited remainder
        // of the tree is never touched.
        if (node == null || state[0] == 0) return;
        inorder(node.left, state);
        state[0]--;
        if (state[0] == 0) {
            state[1] = node.val;
            return;
        }
        inorder(node.right, state);
    }
}
